### ISSUE TITLE
Commands related to handling kubernetes certificates, broken after migrating the PKI to external CA

### ISSUE THREAD

**Issue by alwaysastudent** [kubernetes/kubeadm#2618]
## What keywords did you search in kubeadm issues before filing this one?

If you have found any duplicates, you should instead reply there and close this page.

If you have not found any duplicates, delete this section and continue on.

## Is this a BUG REPORT or FEATURE REQUEST?

**BUG REPORT**

## Versions

**kubeadm version** (use `kubeadm version`):

```
# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.18", GitCommit:"6f6ce59dc8fefde25a3ba0ef0047f4ec6662ef24", GitTreeState:"clean", BuildDate:"2021-04-15T03:29:14Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Kubernetes version** (use `kubectl version`):
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.18", GitCommit:"6f6ce59dc8fefde25a3ba0ef0047f4ec6662ef24", GitTreeState:"clean", BuildDate:"2021-04-15T03:31:30Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.18", GitCommit:"6f6ce59dc8fefde25a3ba0ef0047f4ec6662ef24", GitTreeState:"clean", BuildDate:"2021-04-15T03:23:41Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
- **Cloud provider or hardware configuration**: 
Not relying on any cloud providers. 

- **OS** (e.g. from /etc/os-release):
```
# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
VERSION_CODENAME=stretch
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

- **Kernel** (e.g. `uname -a`):
```
Linux kernel 5.4.56
```

- **Others**:

## What happened?

We are in the middle of migrating existing kubeadm provisioned k8s PKI, on our cluster, to an external CA (pki). 

All the CA and the certificates are rotated and reissued as part of this effort. We don't lay any private keys for the intermediate CA on the filesystem, and we issue all the certificates, for the k8s cluster, from the upstream PKI infra (which in our case is vault).

The rotation seems to have gone successfully, we have not tested for [adding and removing new nodes](https://github.com/kubernetes/kubeadm/issues/2296) or other complex use-cases yet. But our services on this prototype cluster look good so far.

But one of the things that we noticed is when we perform kubeadm commands related to handling kubernetes certificates, to be specific we would like to check expiry using `kubeadm alpha certs check-expiration` and it fails.

```
# kubeadm alpha certs check-expiration  --v=5
I1206 20:18:25.199366 3649540 cert_rotation.go:137] Starting client certificate rotation controller
[check-expiration] Reading configuration from the cluster...
[check-expiration] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'

open /etc/kubernetes/pki/ca.key: no such file or directory
couldn't load the private key file /etc/kubernetes/pki/ca.key
k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil.TryLoadKeyFromDisk
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil/pki_helpers.go:274
k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil.TryLoadCertAndKeyFromDisk
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil/pki_helpers.go:234
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs.LoadCertificateAuthority
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/certs.go:165
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal.(*kubeConfigReadWriter).Read
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal/readwriter.go:143
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal.(*Manager).GetCertificateExpirationInfo
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal/manager.go:355
k8s.io/kubernetes/cmd/kubeadm/app/cmd/alpha.newCmdCertsExpiration.func1
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/alpha/certs.go:329
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:826
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:914
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:864
k8s.io/kubernetes/cmd/kubeadm/app.Run
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
        _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
        /usr/local/go/src/runtime/proc.go:203
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1357
failed to load key
k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil.TryLoadCertAndKeyFromDisk
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil/pki_helpers.go:236
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs.LoadCertificateAuthority
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/certs.go:165
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal.(*kubeConfigReadWriter).Read
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal/readwriter.go:143
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal.(*Manager).GetCertificateExpirationInfo
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal/manager.go:355
k8s.io/kubernetes/cmd/kubeadm/app/cmd/alpha.newCmdCertsExpiration.func1
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/alpha/certs.go:329
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:826
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:914
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:864
k8s.io/kubernetes/cmd/kubeadm/app.Run
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
        _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
        /usr/local/go/src/runtime/proc.go:203
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1357
failure loading ca certificate authority
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs.LoadCertificateAuthority
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/certs.go:167
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal.(*kubeConfigReadWriter).Read
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal/readwriter.go:143
k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal.(*Manager).GetCertificateExpirationInfo
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal/manager.go:355
k8s.io/kubernetes/cmd/kubeadm/app/cmd/alpha.newCmdCertsExpiration.func1
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/alpha/certs.go:329
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:826
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:914
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:864
k8s.io/kubernetes/cmd/kubeadm/app.Run
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
        _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
        /usr/local/go/src/runtime/proc.go:203
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1357
```

 
Following are the certificates laid on the CP host's filesystem

```
# find /etc/kubernetes/pki/ -type f | sort -nr
/etc/kubernetes/pki/scheduler.key
/etc/kubernetes/pki/scheduler.crt
/etc/kubernetes/pki/sa.pub
/etc/kubernetes/pki/sa.key
/etc/kubernetes/pki/front-proxy-client.key
/etc/kubernetes/pki/front-proxy-client.crt
/etc/kubernetes/pki/front-proxy-ca.crt
/etc/kubernetes/pki/etcd/server.key
/etc/kubernetes/pki/etcd/server.crt
/etc/kubernetes/pki/etcd/peer.key
/etc/kubernetes/pki/etcd/peer.crt
/etc/kubernetes/pki/etcd/healthcheck-client.key
/etc/kubernetes/pki/etcd/healthcheck-client.crt
/etc/kubernetes/pki/etcd/ca.crt
/etc/kubernetes/pki/controller-manager.key
/etc/kubernetes/pki/controller-manager.crt
/etc/kubernetes/pki/ca.crt
/etc/kubernetes/pki/apiserver-kubelet-client.key
/etc/kubernetes/pki/apiserver-kubelet-client.crt
/etc/kubernetes/pki/apiserver.key
/etc/kubernetes/pki/apiserver-etcd-client.key
/etc/kubernetes/pki/apiserver-etcd-client.crt
/etc/kubernetes/pki/apiserver.crt
/etc/kubernetes/pki/admin.key
/etc/kubernetes/pki/admin.crt
```
 

```
root@n231-029-107:~# find /var/lib/kubelet/pki/. | sort -nr
/var/lib/kubelet/pki/./kubelet.key
/var/lib/kubelet/pki/./kubelet.crt
/var/lib/kubelet/pki/./kubelet-client.pem
/var/lib/kubelet/pki/./kubelet-client.key
```

## What you expected to happen?

`kubeadm alpha certs check-expiration` should emit output as expected with EXTERNALLY MANAGED, as yes 

## How to reproduce it (as minimally and precisely as possible)?

Create a cluster with using kubeadm generated PKI and rotate the CA manually without the `apiserver.key`, `ca.key` and `etcd/ca.key` and issue all the required certs. Restart all the components gracefully to complete the migration. 

Execute the  `kubeadm alpha certs check-expiration` to see the error.

## Anything else we need to know?

Is this relevant to the issue mentioned in https://github.com/kubernetes/kubeadm/issues/2296 ?

**Comment by neolit123** [comment]
thanks for logging the issue. that is an actual bug and i confirmed it locally.
with a modification to the code it works fine.

the change is relatively small and we might be able to backport it versions 1.20+ after it merges in master.
it cannot be backported to older versions because they are out of support.

https://kubernetes.io/releases/

**Comment by neolit123** [comment]
PR with the fix is here:
https://github.com/kubernetes/kubernetes/pull/106854
not sure when we can merge it as we are about to release 1.23 today.

but once a change lands in release x.y (preferably a version closer to yours - e.g. 1.20) you can take that kubeadm binary and use it instead of your 1.18 binary. NOTE: the new binary should only be used for the "kubeadm certs" command. there is no longer a "kubeadm alpha certs" command and it has graduated to a top level "kubeadm certs".

**Comment by neolit123** [comment]
post fix output:

```
$ sudo kubeadm certs check-expiration --v=5
[check-expiration] Reading configuration from the cluster...
[check-expiration] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
I1207 20:59:56.312554   70986 kubelet.go:91] attempting to download the KubeletConfiguration from the new format location (UnversionedKubeletConfigMap=true)
I1207 20:59:56.669155   70986 kubelet.go:94] attempting to download the KubeletConfiguration from the DEPRECATED location (UnversionedKubeletConfigMap=false)
W1207 20:59:56.686811   70986 utils.go:69] The recommended value for "resolvConf" in "KubeletConfiguration" is: /run/systemd/resolve/resolv.conf; the provided value is: /run/systemd/resolve/resolv.conf
I1207 20:59:56.686909   70986 common.go:117] WARNING: tolerating control plane version v1.22.2, assuming that k8s version 1.23.0 is not released yet

I1207 20:59:56.687006   70986 certs.go:490] validating certificate period for CA certificate
I1207 20:59:56.687345   70986 certs.go:490] validating certificate period for API server certificate
I1207 20:59:56.687551   70986 certs.go:490] validating certificate period for API server kubelet client certificate
I1207 20:59:56.690129   70986 certs.go:490] validating certificate period for front-proxy CA certificate
CERTIFICATE                EXPIRES                  RESIDUAL TIME   CERTIFICATE AUTHORITY   EXTERNALLY MANAGED
admin.conf                 Dec 07, 2022 17:24 UTC   364d                                    yes     
apiserver                  Dec 07, 2022 17:24 UTC   364d            ca                      yes     
apiserver-etcd-client      Dec 07, 2022 17:24 UTC   364d            etcd-ca                 no      
apiserver-kubelet-client   Dec 07, 2022 17:24 UTC   364d            ca                      yes     
controller-manager.conf    Dec 07, 2022 17:24 UTC   364d                                    yes     
etcd-healthcheck-client    Dec 07, 2022 17:24 UTC   364d            etcd-ca                 no      
etcd-peer                  Dec 07, 2022 17:24 UTC   364d            etcd-ca                 no      
etcd-server                Dec 07, 2022 17:24 UTC   364d            etcd-ca                 no      
front-proxy-client         Dec 07, 2022 17:24 UTC   364d            front-proxy-ca          no      
scheduler.conf             Dec 07, 2022 17:24 UTC   364d                                    yes     

CERTIFICATE AUTHORITY   EXPIRES                  RESIDUAL TIME   EXTERNALLY MANAGED
ca                      Dec 05, 2031 17:24 UTC   9y              yes     
etcd-ca                 Dec 05, 2031 17:24 UTC   9y              no      
front-proxy-ca          Dec 05, 2031 17:24 UTC   9y              no      
```

**Comment by alwaysastudent** [comment]
> PR with the fix is here: [kubernetes/kubernetes#106854](https://github.com/kubernetes/kubernetes/pull/106854) not sure when we can merge it as we are about to release 1.23 today.
> 
> but once a change lands in release x.y (preferably a version closer to yours - e.g. 1.20.zz) you can take that kubeadm binary and use it instead of your 1.18 binary. NOTE: the new binary should only be used for the "kubeadm certs" command. there is no longer a "kubeadm alpha certs" command and it has graduated to a top level "kubeadm certs".

Thank you, but unfortunately, we can not upgrade immediately our k8s clusters. But something we will have to do and it makes sense.

I wonder your fix takes care of `etcd-ca` and `front-proxy-ca` as well.

**Comment by neolit123** [comment]
> I wonder your fix takes care of etcd-ca and front-proxy-ca as well.

the problem was only in kubeconfig checks and on disk certs should be fine.

**Comment by pacoxu** [comment]
It sounds like that you can use the latest kubedm(currently you can build one with the master codes) bin to check the certs for an old version Kubernetes cluster.

Can we close this since https://github.com/kubernetes/kubernetes/pull/106854 was merged?

**Comment by neolit123** [comment]
Certainly not a critical bug, but I think we can backport the fix given it
is small change and the bug is command blocking. Building from source is
not ideal for all users.

**Comment by neolit123** [comment]
> I wonder your fix takes care of etcd-ca and front-proxy-ca as well.

i tested deleting the ca.key, front-proxy-ca.key and etcd/ca.key and the command works, but looks like the etcd ca is not considered external. this looks like another bug.

**Comment by neolit123** [comment]
> i tested deleting the ca.key, front-proxy-ca.key and etcd/ca.key and the command works, but looks like the etcd ca is not considered external. this looks like another bug.

here is the fix for that:
https://github.com/kubernetes/kubernetes/pull/106891

**Comment by alwaysastudent** [comment]
@neolit123 - I was able to test the fix, and I think I found another new issue, perhaps I could open a separate one to track this https://github.com/kubernetes/kubeadm/issues/2628
